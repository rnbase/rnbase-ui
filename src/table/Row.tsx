import React from 'react';
import {
  StyleProp,
  StyleSheet,
  Text,
  TextStyle,
  TouchableHighlight,
  View,
  ViewProps,
} from 'react-native';

import { Theme, Themed, withTheme } from '../theming';

interface RowProps extends ViewProps {
  children?: React.ReactNode;
  title?: string;
  subtitle?: string;
  activeOpacity?: number;
  underlayColor?: string;
  titleStyle?: StyleProp<TextStyle>;
  subtitleStyle?: StyleProp<TextStyle>;
  onPress?: () => void;
}

const Row: React.FC<Themed<typeof createStyleSheet, RowProps>> = ({
  theme: { styles, colors },
  children,
  title,
  subtitle,
  activeOpacity = 1,
  underlayColor = colors.underlay,
  style,
  titleStyle,
  subtitleStyle,
  onPress,
  ...props
}) => {
  const titleText = title && (
    <Text key="title" style={[styles.title, titleStyle]} numberOfLines={1}>
      {title}
    </Text>
  );
  const subtitleText = subtitle && (
    <Text key="subtitle" style={[styles.subtitle, subtitleStyle]} numberOfLines={1}>
      {subtitle}
    </Text>
  );

  const label = !subtitle ? (
    titleText
  ) : (
    <View key="label" style={styles.label}>
      {titleText}
      {subtitleText}
    </View>
  );

  const content = (
    <View {...props} style={[styles.root, style]}>
      {label}
      {children}
    </View>
  );

  if (onPress) {
    return (
      <TouchableHighlight
        onPress={onPress}
        activeOpacity={activeOpacity}
        underlayColor={underlayColor}
      >
        {content}
      </TouchableHighlight>
    );
  }

  return content;
};

const createStyleSheet = ({ colors, fonts }: Theme) =>
  StyleSheet.create({
    root: {
      minHeight: 43,
      flexWrap: 'nowrap',
      flexDirection: 'row',
      alignItems: 'center',
      paddingVertical: 4,
      paddingHorizontal: 20,
      justifyContent: 'space-between',
    },
    label: {
      minHeight: 50,
      justifyContent: 'center',
    },
    title: {
      ...fonts.normal,
      fontSize: 17,
      color: colors.black,
    },
    subtitle: {
      ...fonts.normal,
      fontSize: 12,
      marginTop: 3,
      color: colors.gray,
    },
  });

export default withTheme(Row, createStyleSheet, 'Form.Row');
