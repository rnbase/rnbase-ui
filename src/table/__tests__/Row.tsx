import React from 'react';

// Note: test renderer must be required after react-native.
import TestRenderer from 'react-test-renderer';

const TouchableHighlight = (props: any) => React.createElement('TouchableHighlight', props);

jest.doMock(
  'react-native/Libraries/Components/Touchable/TouchableHighlight',
  () => TouchableHighlight,
);

/**
 * Under test
 */
import Row from '../Row';

type Props = React.ComponentProps<typeof Row>;

const Switch = (props: any) => React.createElement('Switch', props);

/**
 * Under test
 */
const createElement = (props: Partial<Props>) => <Row {...props} />;
const createRenderer = (props: Partial<Props>) => TestRenderer.create(createElement(props));

it('should render normally', () => {
  expect(createRenderer({ title: 'title', subtitle: 'subtitle' })).toMatchSnapshot();
});

it('should render with title and detail component', () => {
  expect(createRenderer({ title: 'title', children: <Switch /> })).toMatchSnapshot();
});

it('should render with image source', () => {
  expect(createRenderer({ image: { uri: 'image.png' } })).toMatchSnapshot();
  expect(
    createRenderer({ image: { uri: 'image.png' }, imageStyle: { tintColor: 'red' } }),
  ).toMatchSnapshot();
});

it('should render with image component', () => {
  expect(createRenderer({ image: <Switch /> })).toMatchSnapshot();
  expect(createRenderer({ image: <Switch />, imageStyle: { tintColor: 'red' } })).toMatchSnapshot();
});

it('should render with disclosure indicator', () => {
  expect(createRenderer({ disclosure: undefined, onPress: jest.fn() })).toMatchSnapshot();
  expect(createRenderer({ disclosure: true, onPress: undefined })).toMatchSnapshot();
  expect(createRenderer({ onPress: jest.fn() })).toMatchSnapshot();
});

it('should render without disclosure indicator', () => {
  expect(createRenderer({ onPress: undefined })).toMatchSnapshot();
  expect(createRenderer({ disclosure: false, onPress: jest.fn() })).toMatchSnapshot();
});

it('should render with custom styles', () => {
  expect(
    createRenderer({
      title: 'title',
      titleStyle: { color: 'red' },
      subtitle: 'subtitle',
      subtitleStyle: { color: 'green' },
      image: { uri: 'image.png' },
      imageStyle: { tintColor: 'blue' },
      style: { backgroundColor: 'yellow' },
      activeOpacity: 0.9,
      underlayColor: 'cyan',
      onPress: jest.fn(),
      onPressIn: jest.fn(),
      onPressOut: jest.fn(),
    }),
  ).toMatchSnapshot();
});

it('should handle onPress', () => {
  const onPress = jest.fn();
  const tree = createRenderer({ onPress });

  tree.root.findByType(TouchableHighlight).props.onPress();

  expect(onPress).toBeCalledWith();
});
