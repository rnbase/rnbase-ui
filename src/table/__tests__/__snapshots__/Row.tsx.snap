// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render normally 1`] = `
<View
  style={
    Array [
      Object {
        "minHeight": 58,
      },
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "overflow": "hidden",
        "paddingHorizontal": 20,
        "paddingVertical": 4,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flexGrow": 1,
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "color": "#000000",
            "fontFamily": "System",
            "fontSize": 17,
            "fontWeight": "400",
            "includeFontPadding": false,
            "textAlignVertical": "center",
          },
          undefined,
        ]
      }
    >
      title
    </Text>
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "color": "#838387",
            "fontFamily": "System",
            "fontSize": 12,
            "fontWeight": "400",
            "includeFontPadding": false,
            "marginTop": 3,
            "textAlignVertical": "center",
          },
          undefined,
        ]
      }
    >
      subtitle
    </Text>
  </View>
  <Switch />
</View>
`;

exports[`should render with custom styles 1`] = `
<TouchableHighlight
  activeOpacity={0.9}
  onPress={[MockFunction]}
  onPressIn={[MockFunction]}
  onPressOut={[MockFunction]}
  underlayColor="cyan"
>
  <View
    style={
      Array [
        Object {
          "minHeight": 58,
        },
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "overflow": "hidden",
          "paddingHorizontal": 20,
          "paddingVertical": 4,
        },
        Object {
          "backgroundColor": "yellow",
        },
      ]
    }
  >
    <Image
      source={
        Object {
          "uri": "image.png",
        }
      }
      style={
        Array [
          Object {
            "flexShrink": 0,
            "height": 29,
            "marginRight": 15,
            "width": 29,
          },
          Object {
            "tintColor": "blue",
          },
        ]
      }
    />
    <View
      style={
        Object {
          "flexGrow": 1,
        }
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#000000",
              "fontFamily": "System",
              "fontSize": 17,
              "fontWeight": "400",
              "includeFontPadding": false,
              "textAlignVertical": "center",
            },
            Object {
              "color": "red",
            },
          ]
        }
      >
        title
      </Text>
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#838387",
              "fontFamily": "System",
              "fontSize": 12,
              "fontWeight": "400",
              "includeFontPadding": false,
              "marginTop": 3,
              "textAlignVertical": "center",
            },
            Object {
              "color": "green",
            },
          ]
        }
      >
        subtitle
      </Text>
    </View>
    <Switch />
    <Image
      source={
        Object {
          "testUri": "../../../src/table/assets/disclosure-indicator.png",
        }
      }
      style={
        Object {
          "flexShrink": 0,
          "height": 12,
          "marginLeft": 10,
          "tintColor": "#c7c7cc",
          "width": 8,
        }
      }
    />
  </View>
</TouchableHighlight>
`;

exports[`should render with disclosure indicator 1`] = `
<TouchableHighlight
  activeOpacity={1}
  onPress={[MockFunction]}
  underlayColor="#0001"
>
  <View
    style={
      Array [
        Object {
          "minHeight": 43,
        },
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "overflow": "hidden",
          "paddingHorizontal": 20,
          "paddingVertical": 4,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "flexGrow": 1,
        }
      }
    />
    <Switch />
    <Image
      source={
        Object {
          "testUri": "../../../src/table/assets/disclosure-indicator.png",
        }
      }
      style={
        Object {
          "flexShrink": 0,
          "height": 12,
          "marginLeft": 10,
          "tintColor": "#c7c7cc",
          "width": 8,
        }
      }
    />
  </View>
</TouchableHighlight>
`;

exports[`should render with disclosure indicator 2`] = `
<View
  style={
    Array [
      Object {
        "minHeight": 43,
      },
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "overflow": "hidden",
        "paddingHorizontal": 20,
        "paddingVertical": 4,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flexGrow": 1,
      }
    }
  />
  <Switch />
  <Image
    source={
      Object {
        "testUri": "../../../src/table/assets/disclosure-indicator.png",
      }
    }
    style={
      Object {
        "flexShrink": 0,
        "height": 12,
        "marginLeft": 10,
        "tintColor": "#c7c7cc",
        "width": 8,
      }
    }
  />
</View>
`;

exports[`should render with disclosure indicator 3`] = `
<TouchableHighlight
  activeOpacity={1}
  onPress={[MockFunction]}
  underlayColor="#0001"
>
  <View
    style={
      Array [
        Object {
          "minHeight": 43,
        },
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "overflow": "hidden",
          "paddingHorizontal": 20,
          "paddingVertical": 4,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "flexGrow": 1,
        }
      }
    />
    <Switch />
    <Image
      source={
        Object {
          "testUri": "../../../src/table/assets/disclosure-indicator.png",
        }
      }
      style={
        Object {
          "flexShrink": 0,
          "height": 12,
          "marginLeft": 10,
          "tintColor": "#c7c7cc",
          "width": 8,
        }
      }
    />
  </View>
</TouchableHighlight>
`;

exports[`should render with image 1`] = `
<View
  style={
    Array [
      Object {
        "minHeight": 43,
      },
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "overflow": "hidden",
        "paddingHorizontal": 20,
        "paddingVertical": 4,
      },
      undefined,
    ]
  }
>
  <Image
    source={
      Object {
        "uri": "image.png",
      }
    }
    style={
      Array [
        Object {
          "flexShrink": 0,
          "height": 29,
          "marginRight": 15,
          "width": 29,
        },
        undefined,
      ]
    }
  />
  <View
    style={
      Object {
        "flexGrow": 1,
      }
    }
  />
  <Switch />
</View>
`;

exports[`should render with title only 1`] = `
<View
  style={
    Array [
      Object {
        "minHeight": 43,
      },
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "overflow": "hidden",
        "paddingHorizontal": 20,
        "paddingVertical": 4,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flexGrow": 1,
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "color": "#000000",
            "fontFamily": "System",
            "fontSize": 17,
            "fontWeight": "400",
            "includeFontPadding": false,
            "textAlignVertical": "center",
          },
          undefined,
        ]
      }
    >
      title
    </Text>
  </View>
  <Switch />
</View>
`;

exports[`should render without disclosure indicator 1`] = `
<View
  style={
    Array [
      Object {
        "minHeight": 43,
      },
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "overflow": "hidden",
        "paddingHorizontal": 20,
        "paddingVertical": 4,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "flexGrow": 1,
      }
    }
  />
  <Switch />
</View>
`;

exports[`should render without disclosure indicator 2`] = `
<TouchableHighlight
  activeOpacity={1}
  onPress={[MockFunction]}
  underlayColor="#0001"
>
  <View
    style={
      Array [
        Object {
          "minHeight": 43,
        },
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "overflow": "hidden",
          "paddingHorizontal": 20,
          "paddingVertical": 4,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Object {
          "flexGrow": 1,
        }
      }
    />
    <Switch />
  </View>
</TouchableHighlight>
`;
